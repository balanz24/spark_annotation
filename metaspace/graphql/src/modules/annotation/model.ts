import {
  Entity,
  PrimaryColumn,
  Column, ManyToOne, JoinColumn, OneToMany, Index, PrimaryGeneratedColumn,
} from 'typeorm'
import { MomentValueTransformer } from '../../utils/MomentValueTransformer'
import { Moment } from 'moment'
import { MolecularDB } from '../moldb/model'

@Entity('coloc_job')
export class ColocJob {
  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', name: 'ds_id' })
  datasetId: string;

  @Column({ type: 'int' })
  moldbId: number;

  @Column({ type: 'numeric', precision: 2, scale: 2 })
  fdr: number;

  @Column({ type: 'text' })
  algorithm: string;

  @Column({ type: 'timestamp without time zone', transformer: new MomentValueTransformer() })
  start: Moment;

  @Column({ type: 'timestamp without time zone', transformer: new MomentValueTransformer() })
  finish: Moment;

  @Column({ type: 'text', nullable: true })
  error: string | null;

  @Column({ type: 'int', array: true })
  sampleIonIds: number[];

  @OneToMany(() => ColocAnnotation, colocAnnotation => colocAnnotation.colocJob)
  colocAnnotations: ColocAnnotation[];

  @ManyToOne(() => MolecularDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'moldb_id' })
  molecularDB: MolecularDB;
}

@Entity('coloc_annotation')
export class ColocAnnotation {
  @PrimaryColumn({ type: 'uuid' })
  colocJobId: string;

  /** Index of the annotation's sfAdduct in the parent ColocJob's orderedMols array */
  @PrimaryColumn({ type: 'int' })
  ionId: number;

  @Column({ type: 'int', array: true })
  colocIonIds: number[];

  @Column({ type: 'float4', array: true })
  colocCoeffs: number[];

  @ManyToOne(() => ColocJob, { onDelete: 'CASCADE' })
  @JoinColumn()
  colocJob: ColocJob;
}

@Entity('ion')
@Index(['formula', 'chemMod', 'neutralLoss', 'adduct', 'charge'], { unique: true })
export class Ion {
  @PrimaryGeneratedColumn({ type: 'int' })
  id: number;

  /**
   * All components of the ion in one string, with charge as a '+'/'-' suffix e.g. "C21H41O6P-HO+HO3P-H2O-CO2+Na+".
   * Note that this isn't unique - it's possible for different combinations of chemMods, neutralLosses and adducts
   * to build the same ion.
   */
  @Column({ type: 'text' })
  @Index()
  ion: string;

  /** The formula of the molecule prior to any modifications, e.g. "C21H41O6P".
   *  Internally this has been called "sumFormula" or "sf", but this term is incorrect.
   *  It should be referred to as "formula" or "molecular formula". */
  @Column({ type: 'text' })
  formula: string;

  @Column({ type: 'text', default: '' })
  chemMod: string;

  @Column({ type: 'text', default: '' })
  neutralLoss: string;

  @Column({ type: 'text' })
  adduct: string;

  @Column({ type: 'text', default: '' })
  ionFormula: string;

  /** 1 or -1 */
  @Column({ type: 'smallint' })
  charge: number;
}

export const ANNOTATION_ENTITIES = [
  ColocJob,
  ColocAnnotation,
  Ion,
]
