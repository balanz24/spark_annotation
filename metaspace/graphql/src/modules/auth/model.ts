import 'reflect-metadata'
import { Column, Entity, PrimaryColumn } from 'typeorm'
import { Moment } from 'moment'
import { MomentValueTransformer } from '../../utils/MomentValueTransformer'

@Entity()
export class Credentials {
  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', nullable: true })
  hash: string | null;

  @Column({ type: 'text', nullable: true })
  googleId: string | null;

  @Column({ type: 'text', nullable: true })
  apiKey: string | null;

  @Column({ type: 'timestamp without time zone', nullable: true, transformer: new MomentValueTransformer() })
  apiKeyLastUpdated: Moment | null;

  @Column({ type: 'text', nullable: true })
  emailVerificationToken: string | null;

  @Column({
    type: 'timestamp without time zone',
    nullable: true,
    transformer: new MomentValueTransformer(),
  })
  emailVerificationTokenExpires?: Moment | null;

  @Column({ type: 'boolean', default: false })
  emailVerified: boolean;

  @Column({ type: 'text', nullable: true })
  resetPasswordToken: string | null;

  @Column({
    type: 'timestamp without time zone',
    nullable: true,
    transformer: new MomentValueTransformer(),
  })
  resetPasswordTokenExpires: Moment | null;
}

export const AUTH_ENTITIES = [
  Credentials,
]
