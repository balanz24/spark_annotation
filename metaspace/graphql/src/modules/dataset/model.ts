import { Column, Entity, JoinColumn, ManyToOne, OneToMany, OneToOne, PrimaryColumn } from 'typeorm'
import { Group } from '../group/model'
import { User } from '../user/model'
import { Project } from '../project/model'
import { ExternalLink } from '../project/ExternalLink'
import { DatasetDiagnostic, EngineDataset } from '../engine/model'
import { EnrichmentBootstrap } from '../enrichmentdb/model'

@Entity()
export class Dataset {
  @PrimaryColumn({ type: 'text' })
  id: string;

  @Column({ type: 'uuid' })
  userId: string; // dataset submitter and owner -> edit rights

  @ManyToOne(() => User, user => user.datasets)
  @JoinColumn({ name: 'user_id' })
  user: User;

  @Column({ type: 'text', nullable: true })
  description: string | null;

  @Column({ type: 'uuid', nullable: true })
  groupId: string | null; // dataset belongs to group -> all members have view rights

  @ManyToOne(() => Group)
  @JoinColumn({ name: 'group_id' })
  group: Group;

  @Column({ default: false })
  groupApproved: boolean; // true when submitter is a group member

  // when user manually inputs PI details
  @Column({ type: 'text', nullable: true })
  piName: string | null;

  @Column({ type: 'text', nullable: true })
  piEmail: string | null;

  // sm-engine and sm-graphql create & manage the public.dataset (EngineDataset) and graphql.dataset (Dataset) tables
  // independently, so this relationship doesn't enforce an FK.
  @OneToOne(() => EngineDataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'id' })
  engineDataset: EngineDataset;

  @OneToMany(() => DatasetProject, datasetProject => datasetProject.dataset)
  datasetProjects: DatasetProject[];

  @Column({ type: 'json', nullable: true })
  externalLinks: ExternalLink[] | null;

  @OneToMany(() => DatasetDiagnostic, datasetDiagnostic => datasetDiagnostic.dataset)
  datasetDiagnostics: DatasetDiagnostic[];

  @OneToMany(() => EnrichmentBootstrap, enrichmentBootstrap => enrichmentBootstrap.dataset)
  enrichmentBootstrap: EnrichmentBootstrap[];
}

@Entity({ name: 'dataset_project' })
export class DatasetProject {
  @PrimaryColumn({ type: 'text' })
  datasetId: string;

  @ManyToOne(() => Dataset)
  @JoinColumn({ name: 'dataset_id' })
  dataset: Dataset;

  @PrimaryColumn({ type: 'uuid' })
  projectId: string;

  @ManyToOne(() => Project)
  @JoinColumn({ name: 'project_id' })
  project: Project;

  @Column({ type: 'boolean' })
  approved: boolean;
}

export const DATASET_ENTITIES = [
  Dataset,
  DatasetProject,
]
