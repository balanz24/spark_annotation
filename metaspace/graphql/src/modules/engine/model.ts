import {
  Column,
  Entity,
  Index,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryColumn,
  PrimaryGeneratedColumn,
  Unique,
} from 'typeorm'
import { MomentValueTransformer } from '../../utils/MomentValueTransformer'
import { Ion } from '../annotation/model'
import { MolecularDB } from '../moldb/model'
import { Dataset } from '../dataset/model'
import { Moment } from 'moment'
import { EnrichmentBootstrap } from '../enrichmentdb/model'

export type DatasetStatus = 'QUEUED' | 'ANNOTATING' | 'FINISHED' | 'FAILED';

@Entity({ schema: 'public', name: 'dataset' })
export class EngineDataset {
  @PrimaryColumn({ type: 'text' })
  id: string;
  @Index('ind_dataset_name')
  @Column({ type: 'text', nullable: true })
  name: string | null;
  @Column({ type: 'text', nullable: true })
  inputPath: string | null;
  @Column({ type: 'json', nullable: true })
  metadata: any;
  @Column({ type: 'json', nullable: true })
  roi: any;
  @Column({ type: 'json', nullable: true })
  config: any;
  @Column({ type: 'timestamp without time zone', nullable: true, transformer: new MomentValueTransformer() })
  uploadDt: Date | null;
  @Column({ type: 'text', nullable: true })
  status: DatasetStatus | null;
  @Column({ type: 'timestamp without time zone', transformer: new MomentValueTransformer() })
  statusUpdateDt: Date | null;
  @Column({ type: 'text', nullable: true })
  opticalImage: string | null;
  @Column({ type: 'double precision', array: true, nullable: true })
  transform: number[] | null;
  @Column({ type: 'boolean', default: true })
  isPublic: boolean;
  @Column({ type: 'json', nullable: true })
  acqGeometry: any;
  @Column({ type: 'text', default: 'fs' })
  ionImgStorageType: string;
  @Column({ type: 'text', nullable: true })
  thumbnail: string | null;
  @Column({ type: 'text', nullable: true })
  thumbnailUrl: string | null;
  @Column({ type: 'text', nullable: true })
  ionThumbnail: string | null;
  @Column({ type: 'text', nullable: true })
  ionThumbnailUrl: string | null;
  @Column({ type: 'json', nullable: true })
  sizeHash: any;

  // sm-engine and sm-graphql create & manage the public.dataset (EngineDataset) and graphql.dataset (Dataset) tables
  // independently, so this relationship doesn't enforce an FK.
  @ManyToOne(() => Dataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'id' })
  dataset: Dataset;

  @OneToMany(() => OpticalImage, opticalImage => opticalImage.dataset)
  opticalImages: OpticalImage[];

  @OneToMany(() => Job, job => job.dataset)
  jobs: Job[];

  @OneToMany(() => PerfProfile, pipelineStats => pipelineStats.dataset)
  pipelineStats: PerfProfile[];

  @OneToMany(() => DatasetDiagnostic, datasetDiagnostic => datasetDiagnostic.engineDataset)
  datasetDiagnostics: DatasetDiagnostic[];

  @OneToMany(() => EnrichmentBootstrap, enrichmentBootstrap => enrichmentBootstrap.engineDataset)
  enrichmentBootstrap: EnrichmentBootstrap[];
}

@Entity({ schema: 'public' })
export class OpticalImage {
  @PrimaryColumn({ type: 'text' })
  id: string;
  @Column({ name: 'ds_id' })
  datasetId: string;
  @Column({ type: 'text' })
  type: string;
  @Column({ type: 'real' })
  zoom: number;
  @Column({ type: 'int' })
  width: number;
  @Column({ type: 'int' })
  height: number;
  @Column({ type: 'real', array: true, nullable: true })
  transform: number[][];
  @Column({ type: 'text', nullable: true })
  url: string;

  @ManyToOne(() => EngineDataset, dataset => dataset.opticalImages, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'ds_id' })
  dataset: EngineDataset;
}

@Entity({ schema: 'public' })
export class Job {
  @PrimaryGeneratedColumn()
  id: number;
  @Column({ name: 'moldb_id', type: 'int' })
  moldbId: number;
  @Column({ name: 'ds_id', nullable: true })
  datasetId: string | null;
  @Column({ type: 'text', nullable: true })
  status: DatasetStatus | null;
  @Column({ type: 'timestamp', nullable: true })
  start: Date | null;
  @Column({ type: 'timestamp', nullable: true })
  finish: Date | null;

  @ManyToOne(() => EngineDataset, dataset => dataset.jobs, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'ds_id' })
  dataset: EngineDataset;

  @ManyToOne(() => MolecularDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'moldb_id' })
  molecularDB: MolecularDB;

  @OneToMany(() => Annotation, annotation => annotation.job)
  annotations: Annotation[];

  @OneToMany(() => DatasetDiagnostic, datasetDiagnostic => datasetDiagnostic.job)
  datasetDiagnostics: DatasetDiagnostic[];
}

// Should match the literal in metaspace/engine/sm/engine/annotation/diagnostics.py
export type DiagnosticType = 'TIC' | 'IMZML_METADATA' | 'FDR_RESULTS'
export const DiagnosticTypeOptions: {[k in DiagnosticType]: k} = {
  TIC: 'TIC',
  IMZML_METADATA: 'IMZML_METADATA',
  FDR_RESULTS: 'FDR_RESULTS',
}

export type DiagnosticImageFormat = 'PNG' | 'NPY' | 'JSON' | 'PARQUET'
export interface DiagnosticImage {
  key: string;
  index?: number;
  image_id: string;
  url: string;
  format: DiagnosticImageFormat;
}

@Entity({ schema: 'public' })
// This is the main index for lookup performance. It enforces uniqueness when jobId is non-null, however because
// NULL != NULL in postgres, this index doesn't enforce uniqueness of (datasetId, type) when jobId is NULL.
// (i.e. counterintuitively, this index won't prevent inserting ('2001-01-01_00h00m00s', 'TIC', NULL) even if an
// identical row already exists, because NULL != NULL, so NULL is always considered a unique value)
// Therefore a second index is used for enforcing uniqueness when jobId is null.
@Index(['datasetId', 'type', 'jobId'], { unique: true })
@Index(['datasetId', 'type'], { unique: true, where: 'job_id IS NULL' })
export class DatasetDiagnostic {
  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', enum: Object.values(DiagnosticTypeOptions) })
  type: DiagnosticType;

  @Column({ name: 'ds_id', type: 'text' })
  datasetId: string;

  @Column({ type: 'int', nullable: true })
  jobId: number | null;

  @Column({
    type: 'timestamp without time zone',
    default: () => "(now() at time zone 'utc')",
    transformer: new MomentValueTransformer(),
  })
  updatedDT: Moment;

  @Column({ type: 'json', nullable: true })
  data: any;

  @Column({ type: 'text', nullable: true })
  error: any;

  // images are stored separately in an array with a defined structure so that it's easy for all images to be cleaned
  // up when a dataset is deleted, without needing to understand any of the structure in `data`.
  // All created Image IDs MUST be included in this column. Additionally, they MAY be referenced in `data`
  // (preferably by `key` instead of `image_id`).
  @Column({ type: 'json' })
  images: DiagnosticImage[];

  // This table is a child table of public.dataset, not graphql.dataset, so avoid making an FK
  @ManyToOne(() => Dataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'ds_id' })
  dataset: Dataset;

  @ManyToOne(() => EngineDataset, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'ds_id' })
  engineDataset: EngineDataset;

  @ManyToOne(() => Job, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'job_id' })
  job: Job;
}

interface AnnotationStats {
  chaos: number;
  spatial: number;
  spectral: number;
  msm: number;
  total_iso_ints: number[];
  min_iso_ints: number[];
  max_iso_ints: number[];
}

type OffSampleLabel = 'off' | 'on';
interface AnnotationOffSample {
  prob: number;
  label: OffSampleLabel;
}

@Entity({ schema: 'public' })
@Unique('annotation_annotation_uindex', ['jobId', 'formula', 'chemMod', 'neutralLoss', 'adduct'])
export class Annotation {
  @PrimaryGeneratedColumn()
  id: number;
  @Index('annotation_job_id_index')
  @Column()
  jobId: number;
  @Column({ type: 'text' })
  formula: string;
  @Column({ type: 'text' })
  chemMod: string;
  @Column({ type: 'text' })
  neutralLoss: string;
  @Column({ type: 'text' })
  adduct: string;
  @Column({ type: 'real' })
  msm: number;
  @Column({ type: 'real' })
  fdr: number;
  @Column({ type: 'json' })
  stats: AnnotationStats;
  @Column({ type: 'text', array: true })
  isoImageIds: string[];
  @Column({ type: 'json', nullable: true })
  offSample: AnnotationOffSample | null;
  @Column({ type: 'int', nullable: true })
  ionId: number | null;

  @ManyToOne(() => Job, job => job.annotations, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'job_id' })
  job: Job;

  @ManyToOne(() => Ion, { onDelete: 'SET NULL' })
  @JoinColumn({ name: 'ion_id' })
  ion: Ion | null;
}

@Entity({ schema: 'public' })
export class PerfProfile {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: 'text' })
  taskType: string;

  @Column({ name: 'ds_id', nullable: true })
  datasetId: string | null;

  @Column({ type: 'timestamp' })
  start: Date;

  @Column({ type: 'timestamp', nullable: true })
  finish: Date | null;

  @Column({ type: 'json', nullable: true })
  extraData: any;

  @Column({ type: 'text', nullable: true })
  logs: string | null;

  @Column({ type: 'text', nullable: true })
  error: string | null;

  @ManyToOne(() => EngineDataset, dataset => dataset.pipelineStats, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'ds_id' })
  dataset: EngineDataset;

  @OneToMany(() => PerfProfileEntry, entry => entry.profile)
  entries: PerfProfileEntry[];
}

@Entity({ schema: 'public' })
export class PerfProfileEntry {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: 'int' })
  profileId: number;

  @Column({ type: 'int' })
  sequence: number;

  @Column({ type: 'timestamp' })
  start: Date;

  @Column({ type: 'timestamp' })
  finish: Date;

  @Column({ type: 'text' })
  name: string;

  @Column({ type: 'json', nullable: true })
  extraData: any;

  @ManyToOne(() => PerfProfile, profile => profile.entries, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'profile_id' })
  profile: PerfProfile;
}

@Entity({ schema: 'public' })
export class ScoringModel {
  @PrimaryGeneratedColumn()
  id: number;

  /** `name` is for finding models based on a dataset's `scoring_model` DSConfig value */
  @Index({ unique: true })
  @Column({ type: 'text' })
  name: string;

  /** `type` is hint for how to interpret `params`. sm-engine is responsible for managing the valid values. */
  @Column({ type: 'text' })
  type: string;

  /** `params` is a JSON blob of anything sm-engine needs to load & use the model. It's intentionally schemaless,
   as it's very likely that future model types will need to extend the structure. */
  @Column({ type: 'json' })
  params: any;
}

export const ENGINE_ENTITIES = [
  EngineDataset,
  OpticalImage,
  Job,
  DatasetDiagnostic,
  Annotation,
  PerfProfile,
  PerfProfileEntry,
  ScoringModel,
]
