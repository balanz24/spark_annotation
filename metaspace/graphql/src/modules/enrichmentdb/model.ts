import {
  Column,
  Entity,
  Index,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
} from 'typeorm'
import { MolecularDB } from '../moldb/model'
import { Dataset } from '../dataset/model'
import { EngineDataset } from '../engine/model'
import { MomentValueTransformer } from '../../utils/MomentValueTransformer'

@Entity({ schema: 'public' })
export class EnrichmentDB {
  @PrimaryGeneratedColumn()
  id: number;

  /** `name` is for equivalent enrichment's reference db name */
  @Index({ unique: true })
  @Column({ type: 'text' })
  name: string;
}

@Entity({ schema: 'public' })
export class EnrichmentTerm {
  @PrimaryGeneratedColumn()
  id: number;

  /** `id` is for equivalent enrichment's reference source id */
  @Index({ unique: true })
  @Column({ type: 'text' })
  enrichmentId: string;

  /** `name` is for equivalent enrichment's reference source name */
  @Column({ type: 'text' })
  enrichmentName: string;

  @Column({ type: 'int' })
  enrichmentDbId: number;

  @ManyToOne(() => EnrichmentDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'enrichment_db_id' })
  enrichmentDB: EnrichmentDB;
}

@Entity({ schema: 'public' })
@Index(['molecularDbId', 'formula'])
export class EnrichmentDBMoleculeMapping {
  @PrimaryGeneratedColumn()
  id: number;

  /** molecular db `name` for molecule to map with reference enrichment db */
  @Column({ type: 'text' })
  moleculeEnrichedName: string;

  @Column({ type: 'text' })
  formula: string;

  @Column({ type: 'int' })
  enrichmentTermId: number;

  @Column({ type: 'int' })
  moleculeId: number;

  @Column({ type: 'int' })
  molecularDbId: number;

  @ManyToOne(() => EnrichmentTerm, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'enrichment_term_id' })
  enrichmentTerm: EnrichmentTerm;

  @ManyToOne(() => MolecularDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'molecular_db_id' })
  molecularDB: MolecularDB;
}

@Entity({ schema: 'public' })
export class EnrichmentBootstrap {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: 'int' })
  scenario: number;

  @Column({ type: 'text' })
  formulaAdduct: string;

  @Column({ name: 'dataset_id' })
  datasetId: string;

  @Column({ name: 'annotation_id' })
  annotationId: string;

  @Column({ type: 'numeric', precision: 2, scale: 2 })
  fdr: string;

  @Column({ type: 'int' })
  enrichmentDbMoleculeMappingId: number;

  // This table is a child table of public.dataset, not graphql.dataset, so avoid making an FK
  @ManyToOne(() => Dataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'dataset_id' })
  dataset: Dataset;

  @ManyToOne(() => EngineDataset, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'dataset_id' })
  engineDataset: EngineDataset;

  @ManyToOne(() => EnrichmentDBMoleculeMapping, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'enrichment_db_molecule_mapping_id' })
  enrichmentDBMoleculeMapping: EnrichmentDBMoleculeMapping;
}

@Entity({ schema: 'public' })
export class DatasetEnrichment {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'dataset_id' })
  datasetId: string;

  @Column({ type: 'int' })
  enrichmentDbId: number;

  @Column({ type: 'timestamp without time zone', nullable: true, transformer: new MomentValueTransformer() })
  processingDt: Date | null;

  @Column({ type: 'int' })
  molecularDbId: number;

  // This table is a child table of public.dataset, not graphql.dataset, so avoid making an FK
  @ManyToOne(() => Dataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'dataset_id' })
  dataset: Dataset;

  @ManyToOne(() => MolecularDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'molecular_db_id' })
  molecularDB: MolecularDB;

  @ManyToOne(() => EnrichmentDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'enrichment_db_id' })
  enrichmentDB: EnrichmentDB;
}

export const ENRICHMENT_DB_ENTITIES = [
  EnrichmentDB,
  EnrichmentTerm,
  EnrichmentDBMoleculeMapping,
  EnrichmentBootstrap,
  DatasetEnrichment,
]
