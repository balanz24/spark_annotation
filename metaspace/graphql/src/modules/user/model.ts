import { Entity, PrimaryColumn, Column, OneToOne, JoinColumn, OneToMany } from 'typeorm'
import { Credentials } from '../auth/model'
import { UserGroup } from '../group/model'
import { Dataset } from '../dataset/model'
import { UserProject } from '../project/model'

@Entity()
export class User {
  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', nullable: true })
  name: string | null;

  @Column({ type: 'text', nullable: true })
  email: string | null;

  @Column({ type: 'text', nullable: true })
  notVerifiedEmail: string | null;

  @Column({ type: 'text', default: 'user' })
  role: 'admin' | 'user';

  @Column({ type: 'text' })
  credentialsId: string;

  @OneToOne(() => Credentials)
  @JoinColumn({ name: 'credentials_id' })
  credentials: Credentials;

  @OneToMany(() => Dataset, ds => ds.user)
  datasets: Dataset[];

  @OneToMany(() => UserGroup, userGroup => userGroup.user)
  groups?: UserGroup[];

  @OneToMany(() => UserProject, userProject => userProject.user)
  projects?: UserProject[];
}

export const USER_ENTITIES = [
  User,
]
