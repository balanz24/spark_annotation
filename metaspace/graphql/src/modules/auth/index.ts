export { configureAuth } from './controller'
export { sendInvitationEmail } from './email'
