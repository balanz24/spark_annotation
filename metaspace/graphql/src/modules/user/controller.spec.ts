import * as _ from 'lodash'
import {
  createTestGroup,
  createTestProject, createTestUser,
  createTestUserGroup,
  createTestUserProject,
} from '../../tests/testDataCreation'
import { UserGroupRole, UserProjectRole } from '../../binding'
import {
  UserProjectRoleOptions as UPRO,
} from '../project/model'
import {
  UserGroupRoleOptions as UGRO,
} from '../group/model'
import { Credentials as CredentialsModel } from '../auth/model'
import {
  doQuery,
  onAfterAll,
  onAfterEach,
  onBeforeAll,
  onBeforeEach,
  setupTestUsers,
  shallowFieldsOfSchemaType,
  testEntityManager,
  testUser, userContext,
} from '../../tests/graphqlTestEnvironment'
import { createBackgroundData, validateBackgroundData } from '../../tests/backgroundDataCreation'
import { getContextForTest } from '../../getContext'
import { utc } from 'moment'

describe('modules/user/controller', () => {
  const setupProject = async(projectRole: UserProjectRole | null, isPublic: boolean) => {
    const project = await createTestProject({ isPublic, name: `${projectRole} project` })
    await createTestUserProject(userContext.user.id!, project.id, projectRole)
    return project
  }

  const setupGroup = async(groupRole: UserGroupRole, primary: boolean) => {
    const group = await createTestGroup()
    await createTestUserGroup(userContext.user.id!, group.id, groupRole, primary)
    return group
  }

  beforeAll(onBeforeAll)
  afterAll(onAfterAll)
  beforeEach(async() => {
    await onBeforeEach()
    await setupTestUsers()
  })
  afterEach(onAfterEach)

  describe('Query.currentUser', () => {
    const query = `query {
      currentUser {
        ${shallowFieldsOfSchemaType('User')}
        groups {
          ${shallowFieldsOfSchemaType('UserGroup')}
          group {
            ${shallowFieldsOfSchemaType('Group')}
          }
        }
        primaryGroup {
          ${shallowFieldsOfSchemaType('UserGroup')}
          group {
            ${shallowFieldsOfSchemaType('Group')}
          }
        }
        projects {
          ${shallowFieldsOfSchemaType('UserProject')}
          project {
            ${shallowFieldsOfSchemaType('Project')}
          }
        }
      }
    }`

    describe('should include project in currentUser.projects', () => {
      const ROLE_COMBOS: [UserProjectRole | null, boolean, boolean][] = [
        [null, false, false],
        [UPRO.PENDING, false, false],
        [UPRO.INVITED, false, true],
        [UPRO.MEMBER, false, true],
        [UPRO.MANAGER, false, true],
        [UPRO.REVIEWER, false, true],
        [UPRO.PENDING, true, true],
        [UPRO.INVITED, true, true],
        [UPRO.MEMBER, true, true],
        [UPRO.MANAGER, true, true],
        [UPRO.REVIEWER, true, true],
      ]
      it.each(ROLE_COMBOS)('project role: %s, project.isPublic: %s, expected: %s',
        async(projectRole, isPublic, expected) => {
          // Arrange
          const project = await setupProject(projectRole, isPublic)

          // Act
          const result = await doQuery(query, { }, { context: userContext })

          // Assert
          expect(result.projects).toEqual(!expected
            ? []
            : [
                expect.objectContaining({
                  project: expect.objectContaining({ id: project.id }),
                  role: projectRole,
                }),
              ])
        })
    })

    it('should match snapshot', async() => {
      // Arrange
      await createBackgroundData({ projects: true, users: true })
      await setupProject(UPRO.MEMBER, false)
      await setupGroup(UGRO.MEMBER, true)
      await setupGroup(UGRO.MEMBER, false)

      // Act
      const result = await doQuery(query, { }, { context: userContext })

      // Assert
      const maskedFields = ['id', 'userId', 'projectId', 'groupId', 'email', 'createdDT']
      const cleanData = (value: any, key?: any): any => {
        if (_.isObject(value)) {
          return _.mapValues(value, cleanData)
        } else if (_.isArray(value)) {
          return _.map(value, cleanData)
        } else if (maskedFields.includes(key) && value) {
          return 'MASKED'
        } else {
          return value
        }
      }
      const currentUser = cleanData(result)
      expect(currentUser).toMatchSnapshot()
    })
  })

  describe('Mutation.resetUserApiKey', () => {
    const query = `mutation($userId: ID!, $removeKey: Boolean!) {
      resetUserApiKey(userId: $userId, removeKey: $removeKey) {
        id
        apiKey
      }
    }`

    it('should generate a 12-digit API key when called by the current user', async() => {
      // Arrange
      const backgroundData = await createBackgroundData({ users: true })

      // Act
      const result = await doQuery(query, { userId: testUser.id, removeKey: false }, { context: userContext })

      // Assert
      const updatedCreds = await testEntityManager.findOneOrFail(CredentialsModel, testUser.credentialsId)
      expect(result.id).toBe(testUser.id)
      expect(result.apiKey).toHaveLength(12)
      expect(result.apiKey).toBe(updatedCreds.apiKey)
      expect(updatedCreds.apiKeyLastUpdated).toBeTruthy()
      await validateBackgroundData(backgroundData)
    })

    it('should set the API key to null when called by the current user with removeKey: true', async() => {
      // Arrange
      const backgroundData = await createBackgroundData({ users: true })
      const startTime = utc().subtract(10, 'days')
      await testEntityManager.update(CredentialsModel, testUser.credentialsId, {
        apiKey: 'ExistingApiKey',
        apiKeyLastUpdated: startTime,
      })

      // Act
      const result = await doQuery(query, { userId: testUser.id, removeKey: true }, { context: userContext })

      // Assert
      const updatedCreds = await testEntityManager.findOneOrFail(CredentialsModel, testUser.credentialsId)
      expect(result.id).toBe(testUser.id)
      expect(result.apiKey).toBeNull()
      expect(updatedCreds.apiKey).toBeNull()
      expect(updatedCreds.apiKeyLastUpdated!.diff(startTime, 'days', true)).toBeCloseTo(10)
      await validateBackgroundData(backgroundData)
    })

    it('should fail when called by a different current user', async() => {
      // Arrange
      await testEntityManager.update(CredentialsModel, testUser.credentialsId, { apiKey: 'ExistingApiKey' })
      const otherUserContext = getContextForTest((await createTestUser()) as any, testEntityManager)

      // Act
      const resultPromise = doQuery(query, { userId: testUser.id, removeKey: false }, { context: otherUserContext })
      await expect(resultPromise).rejects.toThrow()

      // Assert
      const updatedCreds = await testEntityManager.findOneOrFail(CredentialsModel, testUser.credentialsId)
      expect(updatedCreds.apiKey).toBe('ExistingApiKey')
    })
  })
})
