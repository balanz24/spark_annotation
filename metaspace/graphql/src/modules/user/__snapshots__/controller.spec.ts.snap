// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/user/controller Query.currentUser should match snapshot 1`] = `
Object {
  "apiKey": null,
  "email": "MASKED",
  "groups": Object {
    "0": Object {
      "group": Object {
        "currentUserRole": "MEMBER",
        "groupDescriptionAsHtml": "",
        "hasPendingRequest": null,
        "id": "MASKED",
        "name": "test group",
        "numDatabases": 0,
        "numMembers": 1,
        "shortName": "tstgrp",
        "urlSlug": null,
      },
      "numDatasets": 0,
      "role": "MEMBER",
    },
    "1": Object {
      "group": Object {
        "currentUserRole": "MEMBER",
        "groupDescriptionAsHtml": "",
        "hasPendingRequest": null,
        "id": "MASKED",
        "name": "test group",
        "numDatabases": 0,
        "numMembers": 1,
        "shortName": "tstgrp",
        "urlSlug": null,
      },
      "numDatasets": 0,
      "role": "MEMBER",
    },
  },
  "id": "MASKED",
  "name": "tester",
  "primaryGroup": Object {
    "group": Object {
      "currentUserRole": "MEMBER",
      "groupDescriptionAsHtml": "",
      "hasPendingRequest": null,
      "id": "MASKED",
      "name": "test group",
      "numDatabases": 0,
      "numMembers": 1,
      "shortName": "tstgrp",
      "urlSlug": null,
    },
    "numDatasets": 0,
    "role": "MEMBER",
  },
  "projects": Object {
    "0": Object {
      "numDatasets": 0,
      "project": Object {
        "createdDT": "MASKED",
        "currentUserRole": "MEMBER",
        "hasPendingRequest": null,
        "id": "MASKED",
        "isPublic": false,
        "latestUploadDT": null,
        "name": "MEMBER project",
        "numDatasets": 0,
        "numMembers": 1,
        "projectDescription": null,
        "publicationStatus": "UNPUBLISHED",
        "publishedDT": null,
        "reviewToken": null,
        "urlSlug": null,
      },
      "role": "MEMBER",
    },
  },
  "role": "user",
}
`;
